### DISCUSSION TITLE
Does Avalonia feature any efficient means for manual translation/transformation of objects? ie. On GPU?

### DISCUSSION THREAD

**Discussion by jonmdev** [AvaloniaUI/Avalonia#14845] [Q&A] [answered]
In Unity's UI Toolkit (which I think is similar to Avalonia in that both are C# based on VisualElement system but with their own rendering pipeline), there exist the following UsageHints which allow VisualElements to be rapidly transformed or translated:

https://docs.unity3d.com/2022.1/Documentation/ScriptReference/UIElements.UsageHints.html

>Marks a VisualElement that changes its transformation often (i.e. position, rotation or scale). When specified, this flag hints the system to optimize rendering of the VisualElement for recurring transformation changes. The VisualElement's vertex transformation will be done by the GPU when possible on the target platform. Please note that the number of VisualElements to which this hint effectively applies can be limited by target platform capabilities. For such platforms, it is recommended to prioritize use of this hint to only the VisualElements with the highest frequency of transformation changes. 

This allowed me to create my own scroll view and animation type objects that were incredibly efficient and smooth even at 120 fps on mobile devices.

I since switched to Maui and have struggled to replicate this as Android and native development offers no such similar optimizations we have access to. Apparently default native "ScrollView" have optimizations, but we cannot manually access these optimizations for our own objects. So moving things around by Translation in native is terribly slow and awful.

If one was to create a manual scrollwindow or animating carousel/view/etc in Avalonia, how would one optimize it to run efficiently even at 120 hz on a mobile device? Is there any similar such optimization as in Unity?

Thanks for any guidance.

**Comment by maxkatz6** [comment]
Not sure if I understand how this Unity optimization exactly works, but I would assume it's not applicable/possible in Avalonia. 

In Avalonia ScollViewer children are layouted on UI thread as visuals, but not vertices. Then each visual executes its draw calls (draw rectangle, draw bitmap...). After that on Render thread (for the most platform, on Browser - it's single threaded) these draw calls are executed and with help of Skia these commands will be transferred to the GPU. Only in the end geometry/vertices are created, but it might be different depending how Skia decides to handle it. 

Possible optimizations might be:
1. Do not render what's not visible on the screen. It's already handled by Avalonia. But layout is still calculated for all items (with exception of virtualized scrollable lists).
2. Apply smooth scrolling only on Render thread. I.e. animating scroll offset on render thread, without doing any layout recalculation. It's not yet implemented, but for simpler scroll viewers it can be done with some experiments, see https://github.com/AvaloniaUI/Avalonia/issues/12977#issuecomment-1733686691

**Reply by jonmdev** [reply]
To clarify further, basically the point is as in Android with the Hardware acceleration:

https://developer.android.com/topic/performance/hardware-accel

You must be able to "freeze" objects to the GPU and then translate them without redrawing them every time. 

In Android we can do this with:
`view.setLayerType(View.LAYER_TYPE_HARDWARE, null)`

Which then saves it on the GPU and only updates if you do something other than Translate or a few basic other operations and it truly changes.

So for example, if you have an object you are translating (movable element of a custom scroll view type system for example) you would apply this to that view, and it would only update when needed. The translation would cause no redraw update on it.

This is essentially the same optimization I described for Unity above.

Any similar optimization in Avalonia?

Hope that makes sense. If not no worries. Thanks.

**Reply by maxkatz6** [reply, marked as answer]
Yes, it makes sense. No, in Avalonia you can't keep view subtrees on the GPU.

Right now, it is possible to pre-render control tree to the RenderTargetBitmap, which then can be reused while rendering. But right now, RenderTargetBitmap is CPU bound, but I can see it working with GPU memory at some point in the future.
